EosPayload/drivers/RadioDriver.py

```diff
@@ -1,7 +1,91 @@
+from EosLib.packet.transmit_header import TransmitHeader
+from EosPayload.lib.driver_base import DriverBase
+from EosPayload.lib.mqtt import Topic
+
+
 class RadioDriver(DriverBase):
+    _thread_queue = PriorityQueue()
+    sequence_number = 0
+
+    # mapping from destination to mqtt topic
+    device_map = {
+        Device.RADIO: Topic.RADIO_TRANSMIT,
+        Device.MISC_RADIO_1: Topic.PING_COMMAND
+    }
+
+    def setup(self) -> None:
+        con = True
+        while con:
+            try:
+                self.port = XBeeDevice("/dev/cu.usbserial-8", 9600)
+                # self.port = XBeeDevice("COM9", 9600)
+                self.port.open()
+                self.remote = RemoteXBeeDevice(self.port, XBee64BitAddress.from_hex_string(
+                    "13A20041CB89AE"))  # on the chip itself there is a number on the top right. It should be 3!
+                con = False
+            except Exception as e:
+                self._logger.error(f"radio port not open: {e}")
+                time.sleep(10)
+
+    @staticmethod
+    def get_device_id() -> Device:
+        return Device.RADIO
+
+    @staticmethod
+    def get_device_name() -> str:
+        return "radio-driver"
+
     def device_read(self, logger: logging.Logger) -> None:
-        return 0
+        # Receives data from radio and sends it to MQTT
+        def data_receive_callback(xbee_message):
+            packet = xbee_message.data  # raw bytearray packet
+            logger.info("Packet received ~~~~~~")
+            logger.info(packet)
+            packet_object = Packet.decode(packet)   # convert packet bytearray to packet object
+            dest = packet_object.data_header.destination  # packet object
+            if dest in self.device_map:  # mapping from device to mqtt topic
+                mqtt_topic = self.device_map[dest]
+                self._mqtt.send(mqtt_topic, packet)
+            else:
+                logger.info("no mqtt destination mapping")
+
+        # Receives data from MQTT and sends it down to ground station according to priority
+        def xbee_send_callback(client, userdata, message):
+            # gets message from MQTT and convert transmit_packet to packet object (look at Thomas code)
+            packet_from_mqtt = Packet.decode(message.payload)
+
+            # append transmit header
+            new_transmit_header = TransmitHeader(self.sequence_number)
+            packet_from_mqtt.transmit_header = new_transmit_header
+
+            # add packet to queue
+            priority = packet_from_mqtt.data_header.priority
+            logger.info(f"Enqueuing packet seq={self.sequence_number}")
+            self._thread_queue.put((priority, datetime.now(), packet_from_mqtt,))
+
+            self.sequence_number = (self.sequence_number + 1) % 256  # sequence number can't exceed 255
+
+        self._mqtt.register_subscriber(Topic.RADIO_TRANSMIT, xbee_send_callback)
+        self.port.add_data_received_callback(data_receive_callback)
```

Comment:
If you move the above function definitions to outside of the device_read function, you can do all of this in setup, which is generally where MQTT stuff should be handled